tests/test_feeds.py

```diff
@@ -0,0 +1,114 @@
+# Copyright (C) 2020 Zeropoint Dynamics
+
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+
+
+# You should have received a copy of the GNU Affero General Public
+# License along with this program.  If not, see
+# <http://www.gnu.org/licenses/>.
+# ======================================================================
+
+from __future__ import absolute_import
+
+import unittest
+
+from os import path
+
+from zelos import Zelos
+from zelos.feeds import FeedLevel
+
+
+DATA_DIR = path.join(path.dirname(path.abspath(__file__)), "data")
+
+
+class TraceTest(unittest.TestCase):
+    def test_feed_level(self):
+        z = Zelos(None)
+
+        feeds = z.internal_engine.feeds
+        self.assertEqual(feeds.get_feed_level(), FeedLevel.SYSCALL)
+        self.assertFalse(feeds.inst_feed_on)
+        self.assertFalse(feeds.api_feed_on)
+        self.assertTrue(feeds.syscall_feed_on)
+
+        feeds.set_feed_level(FeedLevel.API)
+        self.assertFalse(feeds.inst_feed_on)
+        self.assertTrue(feeds.api_feed_on)
+        self.assertTrue(feeds.syscall_feed_on)
+
+        feeds.set_feed_level(FeedLevel.INST)
+        self.assertTrue(feeds.inst_feed_on)
+        self.assertTrue(feeds.api_feed_on)
+        self.assertTrue(feeds.syscall_feed_on)
+
+        feeds.set_feed_level(FeedLevel.NONE)
+        self.assertFalse(feeds.inst_feed_on)
+        self.assertFalse(feeds.api_feed_on)
+        self.assertFalse(feeds.syscall_feed_on)
+
+    def test_syscall_subscribe(self):
+        z = Zelos(
+            path.join(DATA_DIR, "static_elf_helloworld"),
+            syscall_feed="syscall=set_thread_area",
+            stop_feed="syscall=readlink",
+            inst_feed="syscall=fstat64",
+            api_feed="syscall=write",
+        )
+
+        feeds = z.internal_engine.feeds
+
+        syscalls = []
+
+        def syscall_feed_subscriber(zelos, sysname, args, retval):
+            syscalls.append(sysname)
+
+        feeds.subscribe_to_syscall_feed(syscall_feed_subscriber)
+
+        syscalls2 = []
+
+        def syscall_feed_subscriber2(zelos, sysname, args, retval):
+            syscalls2.append(sysname)
+
+        handle = feeds.subscribe_to_syscall_feed(syscall_feed_subscriber2)
+        feeds.unsubscribe_from_feed(handle)
```

Comment:
Perhaps adding an additional case where syscalls are recorded using a syscall hook directly, and comparing with the syscalls from the syscall feed subscription? To make sure that feeds & syscalls are always appropriately synced.

(Could also do this for instructions/apis)